torch/autograd/functional.py

```diff
@@ -429,6 +467,85 @@ def jacobian(func, inputs, create_graph=False, strict=False):
+
+    if vectorize:
+        if strict:
+            raise RuntimeError('torch.autograd.functional.jacobian: `strict=True` '
+                               'and `vectorized=True` are not supported together. '
+                               'Please either set `strict=False` or '
+                               '`vectorize=False`.')
+        # NOTE: [Computing jacobian with vmap and grad for multiple outputs]
+        #
+        # Let's consider f(x) = (x**2, x.sum()) and let x = torch.randn(3).
+        # It turns out we can compute the jacobian of this function with a single
+        # call to autograd.grad by using vmap over the correct grad_outputs.
+        #
+        # Firstly, one way to compute the jacobian is to stack x**2 and x.sum()
+        # into a 4D vector. E.g., use g(x) = torch.stack([x**2, x.sum()])
+        #
+        # To get the first row of the jacobian, we call
+        # >>> autograd.grad(g(x), x, grad_outputs=torch.tensor([1, 0, 0, 0]))
+        # To get the 2nd row of the jacobian, we call
+        # >>> autograd.grad(g(x), x, grad_outputs=torch.tensor([0, 1, 0, 0]))
+        # and so on.
+        #
+        # Using vmap, we can vectorize all 4 of these computations into one by
+        # passing the standard basis for R^4 as the grad_output.
+        # vmap(partial(autograd.grad, g(x), x))(torch.eye(4)).
+        #
+        # Now, how do we compute the jacobian *without stacking the output*?
+        # We can just split the standard basis across the outputs. So to
+        # compute the jacobian of f(x), we'd use
+        # >>> autograd.grad(f(x), x, grad_outputs=_construct_standard_basis_for(...))
+        # The grad_outputs looks like the following:
+        # ( torch.tensor([[1, 0, 0],
+        #                 [0, 1, 0],
+        #                 [0, 0, 1],
+        #                 [0, 0, 0]]),
+        #   torch.tensor([[0],
+        #                 [0],
+        #                 [0],
+        #                 [1]]) )
+        #
+        # But we're not done yet!
+        # >>> vmap(partial(autograd.grad(f(x), x, grad_outputs=...)))
+        # returns a Tensor of shape [4, 3]. We have to remember to split the
+        # jacobian of shape [4, 3] into two:
+        # - one of shape [3, 3] for the first output
+        # - one of shape [   3] for the second output
+
+        # Step 1: Construct grad_outputs by splitting the standard basis
+        output_numels = tuple(output.numel() for output in outputs)
+        grad_outputs = _construct_standard_basis_for(outputs, output_numels)
+        flat_outputs = tuple(output.reshape(-1) for output in outputs)
+
+        # Step 2: Call vmap + autograd.grad
+        def vjp(grad_output):
+            vj = list(_autograd_grad(flat_outputs, inputs, grad_output, create_graph=create_graph))
+            for el_idx, vj_el in enumerate(vj):
+                if vj_el is not None:
+                    continue
+                vj[el_idx] = torch.zeros_like(inputs[el_idx])
```

Comment:
nit: Shall we use `new_zeros` here as well?

Reply:
Is the suggestion `vj[el_idx] = vj[el_idx].new_zeros(inputs[el_idx].shape)`?

Reply:
`vj[el_idx] = inputs[el_idx].new_zeros(inputs[el_idx].shape)` actually as `vj[el_idx]` is None here. But yes.
I started trying to use the `new_*` variants as much as possible to avoid future errors with vmap. But maybe that's not really necessary?

Reply:
Oh, I see. Thanks for the clarification, I wasn't thinking being able to vmap through this.

`inputs[el_idx].new_zeros(inputs[el_idx].shape)` and `torch.zeros_like(inputs[el_idx])` are the same with respect to vmap: if `inputs[el_idx]` is a BatchedTensor, then that information gets propagated by `new_zeros` and `zeros_like`. What doesn't propagate information is `torch.zeros` because `torch.zeros` doesn't accept a tensor.

Reply:
Thanks for the clarification!